fog/ingest/server/src/controller.rs

```diff
@@ -1028,11 +1081,34 @@ where
         ingress_public_key: &CompressedRistrettoPublic,
         state: &mut MutexGuard<IngestControllerState>,
     ) -> Result<IngressPublicKeyStatus, Error> {
-        self.update_enclave_report_cache()?;
+        // Get a report and check that it makes sense with what we think is happening
+        let report = {
+            let report = self.enclave.get_ias_report()?;
+            // Check that key in report data matches ingress_public_key.
+            // If not, then there is some kind of race.
+            let found_key = try_extract_ingress_pubkey_from_report(&report)?;
+            if &found_key == ingress_public_key {
+                report
+            } else {
+                // Hmm, let's try refreshing the enclave cache
+                self.update_enclave_report_cache()?;
+
+                let report = self.enclave.get_ias_report()?;
+                let found_key = try_extract_ingress_pubkey_from_report(&report)?;
+                if &found_key == ingress_public_key {
+                    report
+                } else {
+                    // This means that the caller is wrong about what the
+                    // current ingress public key is, and we don't have anything we can publish.
+                    log::error!(self.logger, "Report doesn't contain the expected public key even after report refresh: {:?} != {:?}", found_key, ingress_public_key);
+                    return Err(Error::PublishReport);
```

Comment:
Making a note that we will want to have enough information to take action in the error message - what is the node operator expected to do to resolve this scenario?

Reply:
i have no idea how this would happen, we would have to investigate to know what is wrong. it would probably mean that like, some cache involved in creating the reports is in a bad state that is hard to reproduce.

if it doesn't resolve itself then maybe we could try killing the server and making a backup node active instead.